server/sonar-server/src/main/java/org/sonar/server/component/ws/GhostProjectsAction.java

```diff
@@ -0,0 +1,123 @@
+  public static final String ACTION = "ghosts";
+  private final static List<String> POSSIBLE_FIELDS = Arrays.asList("uuid", "key", "name", "creationDate");
+
+  private final DbClient dbClient;
+  private final UserSession userSession;
+
+  public GhostProjectsAction(DbClient dbClient, UserSession userSession) {
+    this.dbClient = dbClient;
+    this.userSession = userSession;
+  }
+
+  @Override
+  public void define(WebService.NewController context) {
+    context
+      .createAction(ACTION)
+      .setDescription("List ghost projects.<br /> Requires admin role.")
+      .setResponseExample(Resources.getResource(getClass(), "projects-example-ghosts.json"))
+      .setSince("5.2")
+      .addPagingParams(100)
+      .addFieldsParam(POSSIBLE_FIELDS)
+      .addSearchQuery("sonar", "names", "keys")
+      .setHandler(this);
+  }
+
+  @Override
+  public void handle(Request request, Response response) throws Exception {
+    userSession.checkGlobalPermission(UserRole.ADMIN);
+    DbSession dbSession = dbClient.openSession(false);
+    SearchOptions searchOptions = new SearchOptions()
+      .setPage(request.mandatoryParamAsInt(Param.PAGE),
+        request.mandatoryParamAsInt(Param.PAGE_SIZE));
+    List<String> desiredFields = fieldsToReturn(request.paramAsStrings(Param.FIELDS));
+    String query = request.param(Param.TEXT_QUERY);
+
+    try {
+      long nbOfProjects = dbClient.componentDao().countGhostProjects(dbSession, query);
+      List<ComponentDto> projects = dbClient.componentDao().selectGhostProjects(dbSession, query, searchOptions);
+      JsonWriter json = response.newJsonWriter().beginObject();
+      writeProjects(json, projects, desiredFields);
+      searchOptions.writeJson(json, nbOfProjects);
+      json.endObject().close();
+    } finally {
+      MyBatis.closeQuietly(dbSession);
+    }
+  }
+
+  private void writeProjects(JsonWriter json, List<ComponentDto> projects, List<String> fieldsToReturn) {
+    json.name("projects");
+    json.beginArray();
+    for (ComponentDto project : projects) {
+      json.beginObject();
+      json.prop("uuid", project.uuid());
+      writeIfWished(json, "key", project.key(), fieldsToReturn);
+      writeIfWished(json, "name", project.name(), fieldsToReturn);
+      writeIfWished(json, "creationDate", project.getCreatedAt(), fieldsToReturn);
+      json.endObject();
+    }
+    json.endArray();
+  }
+
+  private void writeIfWished(JsonWriter json, String key, String value, List<String> fieldsToReturn) {
+    if (fieldsToReturn.contains(key)) {
+      json.prop(key, value);
+    }
+  }
+
+  private void writeIfWished(JsonWriter json, String key, Date value, List<String> desiredFields) {
+    if (desiredFields.contains(key)) {
+      json.propDateTime(key, value);
+    }
+  }
+
+  private List<String> fieldsToReturn(@Nullable List<String> desiredFields) {
+    if (desiredFields == null) {
+      return POSSIBLE_FIELDS;
+    }
+
+    return desiredFields;
```

Comment:
does not apply much here since list is supposed to be short, but as a general practice you should use a Set instead of a List here:
* because you are using only for its contains method, contains implementation is more efficient on Set class
* and you will get the additional benefit of unique values in the collection